[Gabriel Tan] iP

gabztcr opened the pull request.

Commit:
Add Gradle support

gabztcr committed:
Add greeting, echo and exit for Level-1

gabztcr committed:
Add new task and show list functionalities for Level-2

gabztcr committed:
Add new Task class and markAsDone functionality for Level-3

gabztcr committed:
Add Deadline, Event and ToDo classes for Level-4

gabztcr committed:
Add automated text UI testing

gabztcr committed:
Update runtest.bat

gabztcr committed:
Remove additional whitespace

gabztcr committed:
Handle exceptions and errors for Level-5

gabztcr committed:
Add delete functionality and ListChange enum for Level-6 and A-Enums

gabztcr committed:
Update input and expected output files

gabztcr committed:
Add functionality to save tasks using code to read/write

gabztcr committed:
Add date time sensitivity to deadlines and events

gabztcr committed:
Merge branch 'branch-Level-8'

gabztcr committed:
Handle DukeExceptions

gabztcr committed:
Abstract out Ui, Parser, Storage and TaskList classes for Duke

gabztcr committed:
Add JUnit tests for Parser and TaskList classes

gabztcr committed:
Add JavaDoc for some methods

gabztcr committed:
Rename taskify method to textToTask for Java compliance

gabztcr committed:
Add Find functionality to search for tasks

gabztcr committed:
Merge branch 'branch-A-CodingStandard'

gabztcr committed:
Merge branch 'branch-Level-9'

gabztcr committed:
Add a few more JavaDoc

Commit:
build.gradle: Update version to 8.29

ChooJiaXin commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,25 @@
+        isDone = true;
+    }
+    
+    public static Task textToTask(String taskString) throws DukeException {
+        return Parser.getTask(taskString);
+    }
```
Coding standards suggest that verbs should be used to name methods. Perhaps changing textToTask() to getTextToTask() instead?

chrisgzf replied:
+1 on Jia Xin's suggestions. Method names should start with a imperative verb.

ChooJiaXin commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,156 @@
+                listChangePrint(newTask, ListChange.ADD);
+            }
+        }
+    }
```
Similar to textToTask() method, may want to consider changing newToDoT() to getNewToDo() or createNewToDo().

ChooJiaXin commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,156 @@
+        } catch (Exception e) {
+            new DukeException("invalidDelete");
+        }
+    }
```
At first glance, I'm not very sure what this method is deleting. Perhaps renaming the method to deleteTask() would make it clearer as to what you are deleting?

ChooJiaXin commented on src/test/java/TaskListTest.java:
```diff
@@ -0,0 +1,30 @@
+import java.util.List;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class TaskListTest {
+    @Test
+    public void numTasksTest1(){
+        List<Task> lst = new ArrayList<>();
+        lst.add(new Task("test"));
+        TaskList testList = new TaskList(lst);
+        
+        String expected = "1 task";
+        String result = testList.numTasks();
+        assertEquals(expected, result);
+    }
+
+    @Test
+    public void numTasksTest2(){
+        List<Task> lst = new ArrayList<>();
+        for (int i = 0; i < 3; ++i) {
+            lst.add(new Task("test"));
+        }
+        TaskList testList = new TaskList(lst);
+
+        String expected = "3 tasks";
```
Naming standards suggest that we should use names to explain and distinguish between methods and variables. I'm not very sure as to what you are trying to test just by looking at the names of the test methods. It may be better to explain what each test does in its name.

ChooJiaXin reviewed (commented):
LGTM. A few nits but otherwise, your code follows the coding naming standards pretty well. 
Good job :)

gabztcr committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

gabztcr committed:
-Add checkstyle files
-Add JavaDoc for the remaining attributes and methods
-Amend all files to adhere to checkstyle

gabztcr committed:
Amend code to store and read binary for task statuses in duke.txt file

chrisgzf commented on build.gradle:
```diff
@@ -0,0 +1,46 @@
+}
+
+application {
+    mainClassName = "seedu.duke.Duke"
```
Not sure what issues this might cause in the future, but it might be safer to set it as your actual package name in the future.

chrisgzf commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,42 @@
+
+public class Event extends Task {
+    /** Start datetime of the Event. */
+    protected LocalDateTime start;
```
Since there are no classes inheriting from Event, should we set it to `private` first?

blackonyyx replied:
To add on, perhaps can consider abstracting Event and Deadline together as a abstract TimedTask.

chrisgzf commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,42 @@
+    protected LocalDateTime start;
+
+    /** End datetime of the Event. */
+    protected LocalDateTime end;
```
Same as above

chrisgzf commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,42 @@
+            } else {
+                throw new DukeException("invalidEventChronology");
+            }
+        } catch (Exception e) {
```
Do you want to catch more specific exceptions? Might be better to catch a `DateTimeParseException` instead.

chrisgzf commented on src/main/java/DukeException.java:
```diff
@@ -0,0 +1,31 @@
+     * @param errorType Description of error.
+     */
+    public DukeException(String errorType) {
+        String errorMessage = errorType.equals("invalidCommand")
```
Might be worth using an enum for `errorType`, because it seems weird that you are comparing strings. Might lead to runtime improvements too.

chrisgzf commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,29 @@
+
+public class Deadline extends Task {
+    /** Deadline of the task. */
+    protected LocalDateTime by;
```
Same as above, should use `private`

chrisgzf commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,29 @@
+
+    @Override
+    public String toString() {
+        String datetime = Parser.getDateTime(by);
```
```suggestion
        String dateTime = Parser.getDateTime(by);
```
We should be consistent and use camelCase for our variables.

chrisgzf commented on src/main/java/ReadFile.java:
```diff
@@ -0,0 +1,59 @@
+import java.io.FileReader;
+import java.io.IOException;
+
+public class ReadFile {
```
I feel the class name shouldn't sound imperative. Perhaps something like `DukeFileReader` (to prevent confusion with `FileReader`) would be better?

chrisgzf commented on src/main/java/ReadFile.java:
```diff
@@ -0,0 +1,59 @@
+            ++numLines;
+        }
+
+        bufferedReader.close();
```
Might be better to wrap this into either a try-with-resources or try-catch-finally, so that the `.close()` will be run regardless of whether an exception occurs or not. In this case, when a IOException occurs in the previous lines, the `bufferedReader.close()` won't actually be executed.

chrisgzf commented on src/main/java/ReadFile.java:
```diff
@@ -0,0 +1,59 @@
+            data[i] = bufferedReader.readLine();
+        }
+
+        bufferedReader.close();
```
Same as above

chrisgzf commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,83 @@
+
+public class Storage {
+    /** Destination path of the duke.txt storage file. */
+    protected String filePath;
```
Same as above

chrisgzf commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,181 @@
+
+public class TaskList {
+    /** The list of tasks present. */
+    protected List<Task> tasks;
```
Same as above

chrisgzf commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,181 @@
+     * @return List of Tasks.
+     */
+    public List<Task> getTasks() {
+        return this.tasks;
```
Let's try to avoid unnecessary use of `this` (from advanced style guide)

chrisgzf commented on src/main/java/Ui.java:
```diff
@@ -0,0 +1,48 @@
+     * Prints the opening lines in Duke.
+     */
+    public void opening() {
+        this.showLine();
```
Same, avoid unnecessary use of `this`

chrisgzf reviewed (commented):
Code looks okay overall! I just felt that the way you pass magical strings around to indicate different errors feels a bit iffy. It might be difficult to maintain as the codebase grows bigger and might be harder for the IDE to perform static analysis. Would recommend using enums for that use case instead. Also, watch out for access modifiers that are too permissive, and unnecessary use of `this`. Overall, a great attempt! 💯🚀🚀

gabztcr committed:
Add basic GUI for Duke

gabztcr committed:
Refactor DukeExceptions, add assertions, and update JavaDocs

gabztcr committed:
Rename methods/variables, add final keywords for code quality.

gabztcr committed:
Merge pull request #1 from gabztcr/branch-A-Assertions

Refactor DukeExceptions, add assertions, and update JavaDocs

gabztcr committed:
Merge branch 'master' into branch-A-CodeQuality

gabztcr committed:
Use streams for TaskList printing, amend find method, use lambdas for stream and app closing

gabztcr committed:
Merge pull request #2 from gabztcr/branch-A-CodeQuality

Merge branch-A-CodeQuality

gabztcr committed:
Merge branch 'master' into branch-A-Streams

gabztcr committed:
Merge pull request #3 from gabztcr/branch-A-Streams

Use streams for TaskList printing, amend find method, use lambdas for…

gabztcr committed:
Add gradle.yml for CI

gabztcr committed:
Add functionality to mark done/delete multiple tasks at once for extension C-MassOps

gabztcr committed:
Amend CI file

blackonyyx commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,42 @@
+import java.time.LocalDateTime;
+
```
Could consider adding javadocs for class headers.

blackonyyx commented on src/main/java/MainWindow.java:
```diff
@@ -0,0 +1,76 @@
+import java.util.concurrent.CompletableFuture;
```
Would be good to optimise the imports by taking out unused imports in the code.

blackonyyx commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,212 @@
+        String idString = command.split(" ", 2)[1];
+        String[] ids = idString.split(",");
+        return Arrays.stream(ids)
+                     .map(x -> Integer.parseInt(x) - 1)
```
As streams are quite hard to read already, would be good to give meaningful names to variables mapped.

blackonyyx commented on src/main/java/MainWindow.java:
```diff
@@ -0,0 +1,76 @@
+
+    /**
+     * Creates two dialog boxes by default, one echoing user input and the other containing Duke's reply
+     * and then appends them to the dialog container. Clears the user input after processing.
```
Could add why there is a use of CompletableFuture. I actually would like to know also haha

blackonyyx reviewed (commented):
Code looks okay overall, follows most of the coding standards, besides that which was pointed out by previous reviewer.

gabztcr committed:
Add new display pictures and edit dialog css

gabztcr committed:
Add Window Title, UI screenshot and User Guide

gabztcr committed:
Rename Ui.png

gabztcr committed:
Amend Ui.png directory path for User Guide
